src/plugins/ui_actions/public/actions/create_action.ts

```diff
@@ -17,25 +17,16 @@
  * under the License.
  */
 
-import { ActionContextMapping } from '../types';
-import { ActionByType } from './action';
-import { ActionType } from '../types';
-import { ActionDefinition } from './action';
+import { ActionDefinition, Action } from './action';
 
-interface ActionDefinitionByType<T extends ActionType>
-  extends Omit<ActionDefinition<ActionContextMapping[T]>, 'id'> {
-  id?: string;
-}
-
-export function createAction<T extends ActionType>(
-  action: ActionDefinitionByType<T>
-): ActionByType<T> {
+export function createAction<Context extends object = object>(
+  action: ActionDefinition<Context>
+): Action<Context> {
   return {
     getIconType: () => undefined,
     order: 0,
-    id: action.type,
```

Comment:
I'm not sure why this wasn't failing before, but `ActionDefintion` currently has `id` marked as a required field, yet this function does provide a default fallback to the `type` field.

TS was complaining here that this line would be overridden by the `...action` spread below which makes sense, so I removed this line. However, that also required me to add the `id` field to many call sites of this function. I didn't want to change the types, but I think the desired behavior is:
- `ActionDefinition` should require either `id` OR `type`
- If `id` is undefined, `createAction` should fallback to `type`. Is that correct?

Reply:
```
However, that also required me to add the id field to many call sites of this function. I didn't want to change the types, but I think the desired behavior is:

ActionDefinition should require either id OR type
If id is undefined, createAction should fallback to type. Is that correct?
```

This seems correct, but to not add `id` in bunch of call sites, why you didn't make `id` optional for this function lit it was before? 
See: https://github.com/elastic/kibana/pull/87075/files?file-filters%5B%5D=.ts&file-filters%5B%5D=.tsx#diff-4960a0b834fad6726535d59e6fc8b2febb31f0669916cbcfe9f19eec00478f46L26

Reply:
in the old implementation this was valid: 
```js
createAction({ execute() {} }) 
```
since both `id` and `type` used to be optional.

Josh decided to make `id` required instead of introducing a union type `ActionDefinitionWithId | ActionDefinitionWithType`